### DISCUSSION TITLE
Protocol upgrade request not supported with Java Client

### DISCUSSION THREAD

**Discussion by Judekeyser** [Kludex/uvicorn#1965] [Potential Issue]
Hello, I've been running an application (framework less!) with GUnicorn and Uvicorn. The set-up goes as the documentation in Uvicorn explains: use the default worker class for GUnicorn, and install Uvicorn using:
```
pip install 'uvicorn[standard]'
```
Everything is actually inside a Docker (which explicitly runs the above line, plus other dependencies) and this is the result of `pip freeze`:
```
anyio==3.6.2
click==8.1.3
gunicorn==20.1.0
h11==0.14.0
httptools==0.5.0
idna==3.4
psycopg==3.1.8
psycopg-c==3.1.8
psycopg-pool==3.1.7
python-dotenv==1.0.0
python-rapidjson==1.10
PyYAML==6.0
sniffio==1.3.0
typing_extensions==4.5.0
uvicorn==0.22.0
uvloop==0.17.0
watchfiles==0.19.0
websockets==11.0.2
```
There is nothing specific in the configuration of GUnicorn (except the worker class `'uvicorn.workers.UvicornWorker'`.

Everything is working really fine and I can use psycopg in asynchronous mode for example. I have no issue when issuing `curl` requests. However, when I attack my server with a Java client, I get those errors:
```
[2023-05-01 15:38:03 +0000] [13] [WARNING] Unsupported upgrade request.
[2023-05-01 15:38:03 +0000] [13] [WARNING] No supported WebSocket library detected. Please use "pip install 'uvicorn[standard]'", or install 'websockets' or 'wsproto' manually.
```

This is confusing me, because that's exactly how I installed the Uvicorn dependency. Moreover, `websockets` clearly is in the pip freeze result. I know the HttpClient of Java can be tweaked to stick to HTTP 1.1, but I haven't done it (actually I don't own the client process). By default, it looks like it tries to upgrade to websocket protocol.

My question is: did anyone already **solved** the issue (so really solved it, without asking to re-install nor downgrade the client explicitly). In my understanding, it should just work out of the box. But maybe I've missed some configuration somewhere.

Any ideas?

**Comment by Judekeyser** [comment]
(I've found many stack overflow threads that relate the same kind of issues (see for example:

- https://stackoverflow.com/questions/66612990/fastapi-python-why-i-get-unsupported-upgrade-request-with-post-request
- https://support.prodi.gy/t/unsupported-upgrade-request/4821
- https://github.com/tiangolo/fastapi/issues/2209

but none of them seem to solve my issue. I can reproduce without Docker

In my understanding, if Uvicorn cannot use WebSockets, it should not accept the upgrade. I'm not quite sure what is exactly failing. I couldn't make my build more "clean", it literally uses the `uvicorn[standard]` (between quotes). I've rescratched the image several times, same issue

The main function also contains
```
    if scope['type'] != 'http':
        raise Exception("Only HTTP protocol is understood")
```
as preliminary check, but still no change

**Comment by Kludex** [comment]
How are you running your server? Command please.

Run `python -m pip freeze`, and run the server with `python -m gunicorn ...`. If this doesn't solve the issue, please provide an MRE.

**Reply by Judekeyser** [reply]
The result of pip freeze is printed above.

I run the server by the using (as pointed out in the gunicorn doc):
```
gunicorn --config=app.conf.py app:create
```
The configuration file is:
```
import multiprocessing


bind = "0.0.0.0:8000"

workers = multiprocessing.cpu_count()
worker_class = 'uvicorn.workers.UvicornWorker'

preload_app = True
```
It is correctly the uvicorn worker therein, with the uvloop (as expected using the `uvicorn[standard]` install, as explained in the documentation.

It's all running in Docker.

**Reply by Kludex** [reply]
Please create a repository with instructions that I can reproduce the issue.

**Reply by Judekeyser** [reply]
Nvm, I've used Hypercorn instead and it went smoothly. Thx anyway
